### ISSUE TITLE
Incorrect calculation of bitangent on back faces

### ISSUE THREAD

**Issue by alastaira** [traggett/UnitySpriteShaders#2]
Thanks for the work you've put into this asset, and your generosity in sharing it. I've been developing a sprite uber-shader myself over the last few months, but am now considering replacing it with yours instead since I prefer the more modular approach you've taken.

Unfortunately, however, I'm finding a problem with the normal-mapping applied to the back face of a flipped sprite. Specifically, when a sprite is flipped horizontally so its back face faces the camera, the lighting appears to be flipped _vertically_. 

This occurs when either using mesh normals or a fixed normal of (0, 0, -1) (without the backface fix).
![fixednormal](https://cloud.githubusercontent.com/assets/901056/21517258/627c5992-ccd5-11e6-84dd-52c9bbe16550.jpg)

If I select the "Fixed Normal Back Render" checkbox, then it correctly flips the back face, but now the front face is flipped instead! 
![fixednormal_fixbackface](https://cloud.githubusercontent.com/assets/901056/21517259/68968e56-ccd5-11e6-821b-be982c64e556.jpg)

Note that the left/right of the lighting is correct in both cases. This leads me to believe that there's a bug somewhere in the calculation of the bitangent in the following section of code, which is not correctly discerning whether the sprite has been flipped or not, however I haven't quite been able to dig through the code far enough yet to verify exactly what values this is returning for different inputs (also, it only returns a half3, so where is tangent.w getting assigned?!):

```
inline half3 calculateSpriteWorldBinormal(half3 normalWorld, half3 tangentWorld, float tangentW)
{
#if defined(_FIXED_NORMALS_BACK_RENDERING)
	//If we're using fixed normals and sprite is facing away from camera, flip tangentW
	float3 zAxis = float3(0.0, 0.0, 1.0);
	float3 modelForward = mul((float3x3)unity_ObjectToWorld, zAxis);
	float3 cameraForward = mul((float3x3)UNITY_MATRIX_VP, zAxis);
	float directionDot = dot(modelForward, cameraForward);
	//Don't worry if directionDot is zero, sprite will be side on to camera so invisible meaning it doesnt matter that tangentW will be zero too 
	tangentW *= sign(directionDot);
#endif // _FIXED_NORMALS_BACK_RENDERING

	return calculateWorldBinormal(normalWorld, tangentWorld, tangentW);
}
```

I'm using Unity 5.5 under Win10 64bit, and appear to be getting this issue on both the pixel-lit and vertex-lit variations. I'm flipping my sprite by setting the x scale of the transform to -1, although I've also just tried flipping the X axis on the sprite renderer itself and that has the same result.

For what it's worth, the approach I had taken in my own shader was to implement a shader_feature within the vertex shader that hardcoded the tangent as follows:
```
#pragma shader_feature NORMALS_FLIPPED
...
// Normals always point towards camera
v.normal = float3(0, 0, -1);
#if defined(NORMALS_FLIPPED)
v.tangent = float4(1, 0, 0, 1);
#else
v.tangent = float4(1, 0, 0, -1);
#endif
```

I then toggled this feature on or off from my player movement code that was responsible for flipping the sprite so that it stayed in sync with the sprite's facing direction, something like:
```
void Flip()
        {
            facingRight = !facingRight;
            Vector3 theScale = transform.localScale;
            theScale.x *= -1;
            transform.localScale = theScale;
            if (facingRight)
                GetComponent<Renderer>().sharedMaterial.DisableKeyword("NORMALS_FLIPPED");
            else
                GetComponent<Renderer>().sharedMaterial.EnableKeyword("NORMALS_FLIPPED");

        }
```

Hardcoding the tangent in this way creates the correct output for me:
![normals_myversion](https://cloud.githubusercontent.com/assets/901056/21517522/ccd88066-ccd7-11e6-91d5-409956ec221c.jpg)

However, I quite like the elegance of your solution being able to calculate whether the camera is looking at the front or back face of the sprite based on the dot product in the shader - just it doesn't seem to work in my case!

I'll continue investigating myself, but wondered if you had any ideas? As far as I'm aware there's nothing unusual about my setup, but it's interesting that nobody else has reported this issue before (perhaps not that many people are using normal-mapped sprites?) Thanks.

**Comment by traggett** [comment]
Hey!
No worries glad they're useful.
I checked the whole "Fixed Normal Back Render" feature and it does seem broken for me as well. I never use it in my own projects to be honest so might've been broken for a while.
Anyways! I just had a poke around and think I've fixed it by firstly cleaning up that code so its easier to follow (I confused myself with the matrix multiplications!) and by flipping the tangent sign when its turned on - it should be multiplied by the negative dot product rather than just the dot product.
I've checked in this fix so see if it works for you, its definitely something that should be handled just by the shader.

ps the tangent.w is defined by Unity automatically if your vertex struct has a Tangent defined in it (see inside VertexInput at the top of SpriteLighting.cginc).
If you look at this https://docs.unity3d.com/Manual/SL-VertexFragmentShaderExamples.html
There's a comment to explain what the tangent means in Unity:
// in Unity tangents are 4D vectors, with the .w component used to
// indicate direction of the bitangent vector.

**Comment by alastaira** [comment]
Thanks for the response, but just tried the latest version and now the "Fixed Normal Back Rendering" toggle doesn't seem to make any difference at all? 
When illuminated by a single directional light pointing straight down, a front-facing sprite appears correctly lit from above, whereas the back-face of a mirrored sprite appears lit from below, and that remains the same whether I use the back-face rendering fix or not?

**Comment by alastaira** [comment]
Just been digging around in the Unity shader code some more and I think the solution may lie in the (undocumented) variable `unity_WorldTransformParams` defined in UnityShaderVariables.cginc. I'm not sure when this was introduced, but the comments state "w is usually 1.0, or -1.0 for odd-negative scaled transforms".

So, in the case of a sprite that's been mirrored by setting the scale to -1, you can work out whether to flip the bitangent by multiplying by unity_WorldTransformParams.w. In my simple case I can do 

```v.tangent = float4(1, 0, 0, -unity_WorldTransformParams.w);```

because my sprites and camera are always directly facing each other, with their local z axes aligned to the world z axis. I'm not sure how to generalise it into your code because you seem to be doing everything in world space - presumeably to allow for the case where the camera/sprite orientation may be configured differently - top-down, side-on etc. Anyway - just thought I'd bring it to your attention to see if you can figure out a use for it!

**Comment by traggett** [comment]
Hey, happy new year :)

Hmm yeah I'm getting confuesd by this now. 
Unity_WorldTransformParams.w does indeed get set to -1 when the model's scale is negative, and when thats the case you do need to flip the binormal so need to multiply the cross between the normal and tangent by unity_WorldTransformParams.w.
This is alongside the vertex's own tangent direction which is stored in the verts tangent.w as either 1 or -1.
I do all this inside the calculateWorldBinormal() function in ShaderShared.cginc, passing in the vertex's tangent.w and multipling the binormal with both that and unity_WorldTransformParams.w..


However it gets more complicated when using a fixed camera space normal and the model is facing away from the camera - then its tangent will be facing the wrong way meaning we also need to flip the binormal - hence me mutilplying the tangentSign by -1 if thats the case.
(inside calculateSpriteWorldBinormal in SpriteLighting.cginc).

The latest shaders apear to work in all situations for me - I've tested a simple quad with fixed normals turned on and culling turned off.
When the quad is facing away from the camera if I turn off 'fixed normals back rendering' then the direction light looks like its coming from the opposite way, but when I turn it on it looks correct.
This also works when the quad has a negative scale.

I don't no why its not working for you though, it might be worth trying it with a fresh material or resetting the material your testing with as it could maybe have gotten confused with shader defines - unity materials tend to accumulate shader variables and defines so can get conufused.

**Comment by alastaira** [comment]
Thanks for your continued investigation and support (and a Happy New Year to you too!) 

I've certainly encountered strange issues with cached results while doing my own shader development in the past, so I've just tried creating a brand new simple test project: It uses the latest sprite shader code from github, with a material created from the pixel lit shader with backface fix on. I've then applied that material to two instances of the same sprite - one mirrored and one not - and put a single directional light in the scene pointing straight down. 

I've attached the project to this post, and a screenshot of what I see (the normal sprite correctly lit from above, the flipped sprite incorrectly lit from below). Is this not the same as what you see? Have I done something silly wrong?! Thanks.

![sprite shader test](https://cloud.githubusercontent.com/assets/901056/21582569/a70babd4-d055-11e6-93f6-e8a79d9292aa.jpg)



[Sprite Shader Test.zip](https://github.com/traggett/UnitySpriteShaders/files/680139/Sprite.Shader.Test.zip)

**Comment by traggett** [comment]
Hey! ok I've had a look and found some glaring errors with the calculateSpriteWorldBinormal function :/
The first error is it should be checking the dot product of the vert to camera and the mesh normal rather than the camera forward vector and mesh normal.
However the second error is when _FIXED_NORMALS_BACK_RENDERING is enabled, the shader requires a mesh normal even though its using a fixed normal to calculate lighting (to know what direction the vert is facing) - I was doing the dot between the fixed normal and the camera forward which makes absolutely no sense.

I've modified the shader to fix these two bugs, I've attached an updated version of your test project to demonstrate it working with quads.
HOWEVER the sprite renderer is still not working, I'm not sure how a sprite renderer generates it mesh, and how or if it generates normals, tangents etc so will have to look into that next.
If your using the shaders for quads or Spine animations then I recomend using negative scale to flip sprites rather than rotating it away from the camera and using _FIXED_NORMALS_BACK_RENDERING.
It's more work for the shader and means you need to pass through mesh normals as well and negative x scale will give the same results.

[Sprite Shader Test.zip](https://github.com/traggett/UnitySpriteShaders/files/680949/Sprite.Shader.Test.zip)

Anyways, will have a look into Unity's sprite renderer class and see what could be causing the error with it.

**Comment by traggett** [comment]
Ok having dug a little into how the Unity SpriteRenderer creates its meshes behind the scenes I don't think its possible to solve the problem in the shaders.
I think Unity generates incorrect tangents for SpriteRenderer meshes when they are flipped.
When you flip a SpriteRenderer, its normal get flipped correctly (so it looks like its getting lit in the same way as when non-flipped). However the tangent doesn't get flipped meaning the resulting binormal will be upside down.
However it seems unity_WorldTransformParams.w doesn't change when a sprite is flipped, nor its tangent.w's.
This means unlike when rendering normal meshes there's no way for the Shader to know it's rendering a flipped SpriteRenderer so can't know to flip the binormal.

Bottom line is I think it might be an issue on Unity's side, I'm suprised the SpriteRenderer generates tangents for its meshes at all to be honest as they don't support normal mapped sprites.

However! As I mentioned before you can create the same effect as a flipped spriteRenderer just by rotating 180 degrees around the Y axis and using mesh normals.

Hope that solution works for you!

**Comment by alastaira** [comment]
Gah. Having fallen down the rabbit hole this far, I'm beginning to remember why I gave up and just hardcoded my tangents to flip based on a shader feature.... :)

I've logged the fact that unity_WorldTransformParams.w does not get populated for sprites as a bug with Unity together with a simple repo project - I'll post back here if I hear anything from their engineering team, as it seems that that is a complete blocking issue :(

Unfortunately, rotation is not an option for us, as we have various animations and child objects positioned relative to the sprite - rotating around 180 degrees would cause all those child objects "behind" the sprite to suddenly appear "in front" of it instead, which is why we really need to just mirror it about the y axis.

**Comment by traggett** [comment]
Yeah the reason why unity_WorldTransformParams.w doesn't change is by design I think - the Sprite mesh doesn't actually get negatively scaled, instead the Sprites generated verts get positioned flipped I think - read the 'Flipping' section here:
https://docs.unity3d.com/Manual/class-SpriteRenderer.html

Asking Unity to make the tangents correct when a SpriteRenderer is flipped would be the correct way of fixing the problem, although yeah they've previously said Unity's own Sprite system is not designed to work with lighting effects like normal maps - part of the reason why I'm using Spine instead. Its a shame Unity can't make thier own stuff work properly though!

Also! To work with SpriteRenderers correctly these shaders should also being using [PerRendererData] for its texture maps, including normal map, specualar / emission maps etc.
Instead of setting data via the material editor, SpriteRenderer set their texture maps via material property blocks and things like color using vertex colors.

I might be tempted to make a seperate set of these shaders for use with the SpriteRenderer as they need to be fundamentaly different to shaders used on normal meshes / Spine animations etc.

I think I probably named this project incorrectly - thier Uber shaders for rendering Alpha blended meshes in Unity, not for Unitys own sprites (although they can work with them with a bit of work).

**Comment by alastaira** [comment]
The description of flipping on that page is exactly why, in our case, we _do_ want to set a negative transform scale rather than use the flipping feature (which causes the sprite to become misaligned with its collider and child game objects)   :) 

The problem is that we're quite invested in other aspects of Unity's sprite system - we have a custom sprite atlas packing policy, for example, which links to an automated workflow for creating normal maps for the sprites themselves. We also use a custom sprite import post-process which does a much better job of creating tight sprite meshes than the default importer, drastically reducing the amount of overdraw and correctly accounts for sprite meshes with "holes" in the middle, say. I'm not sure I'd consider going back to simple quad meshes now!

p.s. Do you _need_ [PerRendererData] to work correctly with sprites? I know that's how the main sprite texture gets passed in, but what's the problem with passing additional texture maps as normal material properties rather than via material property blocks?

"**Uber shaders for rendering Alpha blended meshes in Unity, not for Unitys own sprites (although they can work with them with a bit of work)**" <- catchy title :)

**Comment by traggett** [comment]
Yeah I use TexturePacker to pack my sprites and Spine to create nice meshes to minimise alpha overdraw, would never suggest using quads! TexturePacker can also generate nice meshes for sprites as well.
Anyways totally understand not wanting to change systems. I do recomend contacting Unity about SpriteRenderer tangents though.
I figured a negative X scale on a SpriteRenderer was doing the same things a flip behind the scenes otherwise I have no idea why unity_WorldTransformParams.w isn't being set to -1, maybe it is doing the same thing, it just scales its children by -1 as well?

To be honest don't really know regarding [PerRendererData]! Unity's SpriteRenderer system is pretty undocumented on what happens behind the scenes with its mesh generation.
It definietly tries to set its textures using a MaterialPropertyBlock and warns you about it using other methods... although you're right this does seem to work even when the shader doesnt support them :/
However when an animation plays with a new spritesheet the corresponding normal map texture will have to be changed manually which isn't great. The same goes if you wanted animated emission or have a specular map.

heh yeah not the best title, at the moment at least Unity's made it impossible for any shaders to support Normal maps correctly on their own SpriteRendering system but if they change things I'll fix them up :)

**Comment by traggett** [comment]
Hey! Good news! I've fixed it :)

I realised even though unity_WorldTransformParams.w isn't being set properly, unity_ObjectToWorld has the objects correct scale set so can use that instead. Using that means negative scaled SpriteRenderers have correct binormals now (there's no need to turn on fixed normals / back face adjusting either unless you want to).

Sprite flipping using the Flip flags still doesn't work but thats up to unity ;)

**Comment by alastaira** [comment]
Hey, that's great - will try to download the latest version and give it a whirl.

Incidentally (even though it may no longer be required) I had a response from the Unity QA team to say that they had been able to reproduce unity_WorldTransformParams.w not being set properly for negatively-scaled sprites as a bug (logged as #867023) and had passed it to the relevant engineering team. 

There's obviously no ETA for resolution on that at the moment, but just in case you ever need to know it in the future!

**Comment by traggett** [comment]
Ah nice. Yeah the maths to use the unity_ObjectToWorld matrix instead is pretty negligible to be fair but good to know it's getting fixed.
Be good to try and get them to fix tangents when using the flip flags too!

**Comment by Clement-E-Todd** [comment]
Hello, this issue seems to exist in the latest version of the asset:

![lighting error](https://user-images.githubusercontent.com/2318146/38471415-e1f97276-3b3e-11e8-974e-31bfaa2dcaea.png)

The statue on the right is scaled (-1, 1, 1). As you can see, the light is hitting it correctly horizontally, but it has been flipped vertically. I tied flipping the sprite in a few different ways (scaling the X axis, rotating on the Y axis, toggling flipX in the sprite renderer) but the results were the same in all cases.
